Use ANSI color codes and Unicode characters on Windows as well

MatmaRex opened the pull request:
# Description

Use ANSI color codes and Unicode characters on Windows as well, when possible. See commit messages for details.


## Suggested changelog entry
* Output now has fewer redundant ANSI color codes.
* On Windows, when using the `--colors` option, the output now has more colors.
* On Windows, when using PHP 7.1+, the output now has more Unicode characters.


## Related issues/external references
(none)


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [ ] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

MatmaRex committed:
Common: prepareForOutput(): Use ANSI color codes on Windows as well

Windows' console supports ANSI escape sequences since Windows 10.
https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences

PHP enables this by default since PHP 7.2.
https://www.php.net/manual/en/function.sapi-windows-vt100-support.php
On older PHP versions, the output with --colors will not be correct,
but this is already the case when using this option.

PHP CodeSniffer special-cased Windows in this code in 2014
(bfd095d39895abef6afae575382d58f7ad058232).
This workaround is no longer needed today in 2024.

Note that the --colors option was already supported on Windows.
This only affects inline highlighting in error and debug messages.

jrfnl commented:
@MatmaRex Thanks for the PR.

However, similar to the other PR, I need to ask you to please either provide a reproduction scenario, screenshots or tests to demonstrate what you are trying to solve. Preferably all three of these should be provided with any PR which changes things in the UI.

I can take a guess what you are trying to solve, but my guess may be wrong, so without any of the above, I can't even begin to properly evaluate the PR.

> PHP enables this by default since PHP 7.2. https://www.php.net/manual/en/function.sapi-windows-vt100-support.php

PHPCS 3.x still has a minimum supported PHP version of PHP 5.4, so we cannot rely on features from PHP 7.2.
Please test your changes with PHP < 7.2 and see what breaks (and yes, things will break).

I could tentatively earmark this PR for PHPCS 4.0 , which will have a PHP 7.2 minimum, but that still doesn't negate my remarks above.

jrfnl commented:
@MatmaRex

> The `--no-colors` mode is unaffected by my patch. I believe nothing new breaks.

Given the following simple file:
```php
<?php

    // This is a space indented line.
    $a = 10;
	// This is a tab indented line.
```

And running the following command on the Windows 10 command line:
```bash
phpcs ./phpcs-662.php --report=code --tab-width=4 --no-colors --standard=generic
```
:point_up: Take note of the use of `--report=code` as that's the report which is being adjusted in this PR (amongst other things). Also note that this is in `no-colors` mode.


On PHP 8.4, everything looks swell:
| Before | After |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/b202ed47-7a56-4ae7-8c94-e0379c2728c8) | ![image](https://github.com/user-attachments/assets/90b2309d-9171-45f3-b2f4-abe8c44b8339)  |


Not so much on PHP 7.0....:
| Before | After |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/fb9ed2d1-eaab-44ca-83c0-2e0cf35c0ab3) | ![image](https://github.com/user-attachments/assets/787ece72-79c1-4aae-8f1f-4412eaaec785) |

Which only goes to demonstrate why I insist on tests and will continue to do so...


> If you really want to support these old systems, the only thing to do would be to disable `--colors` entirely. I could write a patch for that, if you promise you won't ask me for any test cases for it. I'm not sure what the point would be since they wouldn't run in CI anyway.

Proper tests _will_ run in CI and should have caught the above. If not, the tests would not be good enough.

MatmaRex commented:
Oh, right, the "visible" characters are still applied in `--no-colors` mode. Thanks for pointing that out.

The encoding issue is interesting – it happens with PHP 7.0, but not with PHP 7.1, which I tested with.

> Proper tests will run in CI and should have caught the above. If not, the tests would not be good enough.

I'm not very familiar with GitHub's CI – can you point out which of the jobs run on Windows?

jrfnl commented:
> Oh, right, the "visible" characters are still applied in `--no-colors` mode. Thanks for pointing that out.
> 
> The encoding issue is interesting – it happens with PHP 7.0, but not with PHP 7.1, which I tested with.

Might just be that PHP added support for the Windows ANSI codes in PHP 7.1 and only added the function related to it in PHP 7.2.

> 
> > Proper tests will run in CI and should have caught the above. If not, the tests would not be good enough.
> 
> I'm not very familiar with GitHub's CI – can you point out which of the jobs run on Windows?

At this time, there are no runs on Windows as there are no tests covering code which has special conditions for Windows. As soon as such tests would be added, the Windows OS would have to be added to the test matrix/workflow. (No worries about that, I can do that).

MatmaRex commented:
> Might just be that PHP added support for the Windows ANSI codes in PHP 7.1 and only added the function related to it in PHP 7.2.

The ANSI codes and the Unicode characters actually appear unrelated. I found a PHP doc page that documents the latter as a change in PHP 7.1. https://www.php.net/manual/en/migration71.windows-support.php

> At this time, there are no runs on Windows as there are no tests covering code which has special conditions for Windows. As soon as such tests would be added, the Windows OS would have to be added to the test matrix/workflow. (No worries about that, I can do that).

I guess I can try adding some. Probably in yet another pull request.

MatmaRex committed:
Common: prepareForOutput(): Use single color code per run of characters

This reduces the size of the output, and lessens weird effects
of the ANSI color code characters on text wrapping, when the
result of this method is used in a sniff error message
(e.g. see LanguageConstructSpacingSniff).

MatmaRex committed:
Common: prepareForOutput(): Use Unicode output on Windows as well on PHP 7.1+

PHP 7.1 fixed problems with printing Unicode characters to Windows
console that forced us to avoid using them.
https://www.php.net/manual/en/migration71.windows-support.php

MatmaRex commented:
I extended the changes here slightly, adding two additional commits (see commit messages for details):
* Update tests, which were added in #663
* Improve the use of control characters in general
* Use the Unicode characters on PHP 7.1+ on Windows, per our discussion above

MatmaRex commented:
Sorry about the noise, I didn't realize at first that PHPUnit 4.8 (used on PHP 5.4) does not support the `@requires PHP < 7.1` syntax. And thanks for enabling automatic test runs for me :)

MatmaRex commented:
@jrfnl I greatly appreciate your reviews so far. I'd love to hear your thoughts on these patches again whenever you have the time. Please do not feel hurried though, I just want to make it clear that I'm still interested in completing these changes!

jrfnl commented:
> @jrfnl I greatly appreciate your reviews so far. I'd love to hear your thoughts on these patches again whenever you have the time. Please do not feel hurried though, I just want to make it clear that I'm still interested in completing these changes!

@MatmRex Sorry for the delay, looking at it again now.

jrfnl reviewed (changes requested):
@MatmaRex Okay, so I have looked at this again.

Aside from the PR being blocked by the lack of tests for the Code report (which _may_ be fixed via a future iteration on #664), I'm still concerned about the side-effects of this PR.

I've left a number of comments inline for you to consider.

I'm also a little concerned about the performance impact of the liberal use of `array_*` functions in this utility function. On modern PHP versions, it appears to be ~25% slower, while on older PHP versions, it appears to be about ~200% slower.
Before: https://3v4l.org/E3bVW/perf
After: https://3v4l.org/7NQUU/perf

Hope this helps.

jrfnl commented:
@MatmaRex Just checking in - do you intend to continue with this PR ? If not, I suggest closing it.
